tfx_addons/pandas_transform/component.py

```diff
@@ -0,0 +1,350 @@
+  preprocessing TensorFlow graph to be prepended to your trained model.  That means
+  that you **MUST** perform the equivalent preprocessing in your serving client or
+  a similar location in your system.
+
+  Args:
+    examples: A TFX input channel containing a dataset artifact
+    schema: A TFX input channel containing a schema artifact
+    statistics: A TFX input channel containing a statistics artifact
+    transformed_examples: A TFX output channel which will be used to output the resulting
+    dataset artifact
+    module_file: A component parameter containing a file path to a Python file which
+    contains the user code, in a function named 'preprocessing_fn'.
+    beam_pipeline_args: A string with the argv options for creating a Beam pipeline.
+    Note that this is a string, not a list.  It will be split on spaces to create
+    a list. If running TFX >= 1.8.0, if beam_pipeline_args are specified they will
+    override the pipeline beam args.
+
+
+  Returns:
+    The resulting dataset artifact after processing by the user code.
+
+  Raises:
+    ImportError - When the module file is not found.
+
+  Example:
+    from tfx_addons.pandas_transform import PandasTransform
+
+    module_file = 'my_module_file.py'
+    %%writefile {module_file} # Assuming a Jupyter notebook, otherwise a file
+        import numpy as np
+
+        def preprocessing_fn(df, schema, statistics):
+        def zscore(x, stats=None):
+            if x.isnull()[0] or stats[x.name]['std_dev'] == 0.0:
+            return x
+            else:
+            y = (x - stats[x.name]['mean']) / stats[x.name]['std_dev']
+            return y.astype(x.dtype) # to maintain type consistency
+            
+        numeric_cols = df.select_dtypes(include=[np.number]).columns
+        return df[numeric_cols].apply(zscore, stats=statistics)
+
+    _beam_pipeline_args = [
+      '--direct_running_mode=in_memory',
+      '--direct_num_workers=1']
+
+    panda = PandasTransform(
+        examples=example_gen.outputs['examples'],
+        schema=schema_gen.outputs['schema'],
+        statistics=statistics_gen.outputs['statistics'],
+        module_file = os.path.abspath(module_file),
+        beam_pipeline_args=' '.join(_beam_pipeline_args)
+        )
+    context.run(panda, enable_cache=False) # Assuming interactive context
+  """
+
+
+def DoPandasTransform(
+    transformed_examples: tfx.dsl.components.OutputArtifact[Examples],
+    examples: tfx.dsl.components.InputArtifact[Examples],
+    schema: tfx.dsl.components.InputArtifact[Schema],
+    statistics: tfx.dsl.components.InputArtifact[ExampleStatistics],
+    module_file: tfx.dsl.components.Parameter[str],
+    beam_pipeline: beam.pipeline.Pipeline):
+  """The function where the actual transforms are done, for both signatures."""
+  if not os.path.exists(module_file):
+    raise ImportError(
+        'DoPandasTransform: Module file not found: {}'.format(module_file))
```

Comment:
nit: Should we try to check if its a GCS file and download it if so?

Reply:
It looks like that requires some parsing of the URI to extract the bucket name.  I'm not finding any nice clean way to do that.

Reply:
to be fair, we use https://github.com/tensorflow/tfx/blob/b92695748e4ea35164f7791fc3abed0d1d4679e2/tfx/utils/io_utils.py#L38 for some internal use cases so it may work for this as well

Reply:
Good idea, I sorta forgot about that whole thing.